crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+        id: u128,
+        status: RequestStatus,
+        error_message: Option<&str>,
+    ) -> Result<(), StorageError> {
+        let Some(mut existing) = self.get_item_by_id(id).await? else {
+            return Err(StorageError::NotFound(id.to_string()));
+        };
+        existing.status = status;
+        existing.error_message = error_message.map(|s| s.to_string());
+        let payload = serde_json::to_string(&existing).map_err(|e| {
+            StorageError::Internal(format!("Failed to serialize updated request: {e}"))
+        })?;
+
+        let mut builder = self
+            .client
+            .put_item()
+            .table_name(TABLE_NAME)
+            .condition_expression("attribute_exists(id)")
+            .item("id", AttributeValue::S(existing.id.to_string()))
+            .item(
+                "status",
+                AttributeValue::S(status_to_str(&existing.status).to_string()),
+            )
+            .item(
+                "relay_after",
+                AttributeValue::N(existing.relay_after.timestamp().to_string()),
+            )
+            .item(
+                "created_at",
+                AttributeValue::N(existing.created_at.timestamp().to_string()),
+            )
+            .item(
+                "last_note_index",
+                AttributeValue::S(existing.last_note_index.to_string()),
+            )
+            .item("payload", AttributeValue::S(payload));
+
+        if let Some(err) = &existing.error_message {
+            builder = builder.item("error_message", AttributeValue::S(err.clone()));
+        }
+        if let Some(processed_at) = existing.processed_at {
+            builder = builder.item(
+                "processed_at",
+                AttributeValue::N(processed_at.timestamp().to_string()),
+            );
+        }
+
+        match builder.send().await {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                if is_conditional_check_failed(&e) {
+                    Err(StorageError::NotFound(id.to_string()))
+                } else {
+                    Err(map_internal("put_item", e))
+                }
+            }
+        }
+    }
+
+    async fn update_retry_attempt(
+        &self,
+        id: u128,
+        new_relay_after: DateTime<Utc>,
+        new_retry_count: i32,
+        new_error_message: Option<&str>,
+    ) -> Result<(), StorageError> {
+        let Some(mut existing) = self.get_item_by_id(id).await? else {
+            return Err(StorageError::NotFound(id.to_string()));
+        };
+        existing.relay_after = new_relay_after;
+        existing.retry_count = new_retry_count;
+        existing.error_message = new_error_message.map(|s| s.to_string());
+        self.put_full_request(&existing).await
+    }
+
+    async fn get_request_by_last_note_index(
+        &self,
+        last_note_index: &str,
+    ) -> Result<Option<ScheduledRequest>, StorageError> {
+        let out = self
+            .client
+            .scan()
```

Comment:
I think we don't want `scan`, as think we read every item in the table. Instead, we should use GSI I think. This would be better cost-effective.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html